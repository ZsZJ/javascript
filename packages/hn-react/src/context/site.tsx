import { Site } from 'hn';
import React, { createContext, ReactNode } from 'react';
import globalSite from '../utils/site';

// Create a new context.
const { Provider, Consumer } = createContext(globalSite);

// The site provider is the same as the 'Provider', but changes 'site' to
// 'value'.
export const SiteProvider = ({
  site,
  children,
}: {
  site: Site;
  children: ReactNode;
}) => <Provider value={site} children={children} />;

// The consumer is exported as both Site and SiteConsumer. The SiteConsumer can
// be used if also the Site from 'hn' is imported, so the names don't overlap.
export { Consumer as Site, Consumer as SiteConsumer };

// There is also a HOC withSite that adds the site to the props of the
// component. Source of the HOC for Typescript:
// https://stackoverflow.com/a/50613946/1907875
export interface InjectedSiteProps {
  site: Site;
}

export function withSite<P extends InjectedSiteProps>(
  Component: React.ComponentType<P>,
) {
  return function ComponentWithSite(
    props: Pick<P, Exclude<keyof P, keyof InjectedSiteProps>>,
  ) {
    return (
      <Consumer>{theme => <Component {...props} site={theme} />}</Consumer>
    );
  };
}
