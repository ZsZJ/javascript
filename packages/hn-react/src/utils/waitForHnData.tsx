import * as PropTypes from 'prop-types';
import * as React from 'react';

const asyncBootstrapper = require('react-async-bootstrapper'); // tslint:disable-line:no-var-requires

export default async function(children) {
  const context = {
    state: {
      drupalPage: null,
      entities: [],
    },
  };

  class DrupalPageContextProvider extends React.Component {
    static childContextTypes = {
      hnContext: PropTypes.object,
    };

    getChildContext() {
      return { hnContext: context };
    }

    render() {
      return React.Children.only(this.props.children);
    }
  }

  const drupalContextProvider = (
    <DrupalPageContextProvider>{children}</DrupalPageContextProvider>
  );

  await asyncBootstrapper(drupalContextProvider, {
    componentWillUnmount: true,
  });

  return drupalContextProvider;
}
