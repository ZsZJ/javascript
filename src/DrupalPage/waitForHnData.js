import React, { Component } from 'react';
import asyncBootstrapper from 'react-async-bootstrapper'
import PropTypes from 'prop-types';

export default async function (children) {

  const context = {};

  class DrupalPageContextProvider extends Component {
    getChildContext() {
      return {hnContext: context};
    }

    render() {
      return React.Children.only(this.props.children);
    }
  }

  DrupalPageContextProvider.childContextTypes = {
    hnContext: PropTypes.object,
  };

  const drupalContextProvider = (
    <DrupalPageContextProvider>
      {children}
    </DrupalPageContextProvider>
  );

  await asyncBootstrapper(drupalContextProvider);

  return drupalContextProvider;

};
