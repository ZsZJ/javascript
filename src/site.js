import { Site } from 'hn';
import getNested from 'get-nested';

const site = new Site({ url: '' });

if(getNested(() => window.__NEXT_DATA__.HN_DATA)) {
  site.data = window.__NEXT_DATA__.HN_DATA;
  site.url = window.__NEXT_DATA__.HN_URL || site.url;
  site.pagesLoading = {
    [window.location.pathname]: Promise.resolve(),
  };
}

export default site;
