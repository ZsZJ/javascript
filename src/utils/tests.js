import React from 'react';

const mapperComponent = jest.fn((props) => {
  return <div className="MyCustomMapperComponent" {...props} />
});

export const mapper = {
  unique_type__unique_bundle: mapperComponent
};

export const uuid = 'unique-uuid';

export const entity = {
  __hn: {
    entity: {
      type: 'unique_type',
      bundle: 'unique_bundle'
    }
  }
};

export function mockSite() {
  class SiteMock {
    getData = jest.fn(() => {
      return entity;
    });
    getPage = jest.fn(async () => {
      return uuid;
    })
  }
  return new SiteMock();
}
