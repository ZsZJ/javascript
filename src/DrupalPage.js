import React, { Component } from 'react';
import { parse } from 'url';
import PropTypes from 'prop-types';
import getNested from 'get-nested';
import site from './site';

export default class extends Component {
  static async getInitialProps({ asPath }) {
    const location = parse(asPath, true);
    const page = await site.getPage(asPath, true);

    return {
      location,
      page
    };
  }

  static propTypes = {
    layout: PropTypes.oneOfType([
      PropTypes.func,
      PropTypes.string,
    ]),
    mapper: PropTypes.oneOfType([
      PropTypes.shape(),
      PropTypes.func,
    ]).isRequired,
  };

  static defaultProps = {
    layout: 'div',
  };

  render() {
    const mapper = this.props.mapper;

    const Layout = this.props.layout;

    const page = site.getData(this.props.page);

    if(typeof getNested(() => page.type.target_id) !== 'string') {
      return null;
    }

    const ContentType = (typeof mapper === 'object') ? mapper[page.type.target_id] : mapper(page.type.target_id);

    if(!ContentType) {
      console.error('Component for content type', page.type.target_id, 'not found.');
      return null;
    }


    return (
      <Layout
        page={page}
        location={this.props.location}
        history={this.props.history}
      >
        <ContentType page={page} />
      </Layout>
    );
  }
}
