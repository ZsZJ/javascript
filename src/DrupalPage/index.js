import React, { Fragment, Component } from 'react';
import { parse } from 'url';
import PropTypes from 'prop-types';
import getNested from 'get-nested';
import site from '../site';
import EntityMapper from '../EntityMapper';

class DrupalPage extends Component {
  static contextTypes = {
    hnContext: PropTypes.object,
  };

  state = {
    ready: false,
    loadingData: true,
    dataUrl: null,
    pageUuid: null,
  };

  /**
   * If this component exists in a tree that is invoked with the waitForHnData function, this function is invoked.
   * Only after the promise is resolved, the component will be mounted. To keep the data fetched here, we assign the
   * state to the hnContext provided by the DrupalPageContextProvider. This way, the state will be preserved trough
   * multiple renders.
   */
  async asyncBootstrap() {
    await this.loadData(this.props);
    this.context.hnContext.state = this.state;
  }

  /**
   * The first time this element is rendered, we always make sure the component and the Drupal page is loaded.
   */
  componentWillMount() {
    const state = getNested(() => this.context.hnContext.state);
    if (state) {
      this.setState(state);
    } else {
      this.loadData(this.props);
    }
  }

  /**
   * As soon as the url, mapper or asyncMapper props change, we want to load new data.
   * This always unmounts and mounts all children (Layout and ContentType).
   */
  componentWillReceiveProps(nextProps) {
    if(this.props.url !== nextProps.url || this.props.mapper !== nextProps.mapper || this.props.asyncMapper !== nextProps.asyncMapper) {
      this.loadData(nextProps);
    }
  }

  /**
   * This makes sure the data for this url is ready to be rendered.
   * @param url
   * @param mapper
   * @param asyncMapper
   * @returns {Promise.<{pageUuid: string}>}
   */
  static async assureData({ url, mapper, asyncMapper }) {
    // Get the page. If the page was already fetched before, this should be instant.
    const pageUuid = await site.getPage(url);
    if(!pageUuid) {
      throw Error('An error occurred getting a response from the server.');
    }

    return { pageUuid };
  }

  componentWillUnmount() {
    this.lastRequest = null;
  };

  lastRequest = null;

  async loadData({ url, mapper, asyncMapper }) {

    const lastRequest = Symbol(url);

    this.lastRequest = lastRequest;

    this.setState({ loadingData: true });

    // Load the data.
    const { pageUuid } = await DrupalPage.assureData({ url, mapper, asyncMapper});

    // Check if this is still the last request.
    if(this.lastRequest !== lastRequest) return;

    // Mark this component as ready. This mounts the Layout and new ContentType.
    this.setState({ pageUuid, loadingData: false, dataUrl: url });
  }

  render() {
    // Mark this component as not-ready. This unmounts the Layout and old ContentType.
    // Only render if the component is ready.
    if (this.entity && !this.props.renderWhileLoadingData && !this.entity.isReady()) return null;

    // Get props.
    const Layout = this.props.layout;

    // Get the data and content types with the state properties.
    const data = site.getData(this.state.pageUuid);

    return (
      <Layout
        loadingData={this.state.loadingData}
        url={this.state.dataUrl}
        page={data}
        {...this.props.layoutProps}
      >
        <EntityMapper
          mapper={this.props.mapper}
          uuid={this.state.pageUuid}
          asyncMapper={this.props.asyncMapper}
          entityProps={{ ...this.props.pageProps, page: data }}
          ref={(c) => { this.entity = c; }}
        />
      </Layout>
    );
  }
}

DrupalPage.propTypes = {
  url: PropTypes.string.isRequired,
  layout: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.string,
  ]),
  mapper: PropTypes.oneOfType([
    PropTypes.shape(),
    PropTypes.func,
  ]).isRequired,
  asyncMapper: PropTypes.bool,
  layoutProps: PropTypes.shape(),
  renderWhileLoadingData: PropTypes.bool,
  pageProps: PropTypes.shape(),
};

DrupalPage.defaultProps = {
  layout: Fragment,
  asyncMapper: undefined,
  layoutProps: {},
  renderWhileLoadingData: false,
  pageProps: undefined,
};

export default DrupalPage;
