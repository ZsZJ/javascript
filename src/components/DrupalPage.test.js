import DrupalPage from './DrupalPage';
import React from 'react';
import renderer from 'react-test-renderer';
import site from '../utils/site';
import waitForHnData from '../utils/waitForHnData';
import { mapper } from '../utils/tests';

jest.mock('../utils/site', () => {
  return require('../utils/tests').mockSite();
});

jest.mock('util-deprecate', () => jest.fn(func => func));
console.log = console.warn = console.error = jest.fn(message => {
  throw new Error(message);
});

beforeEach(() => {
  site.getData.mockRestore();
});

describe('DrupalPage', async () => {
  test('with required props', async () => {
    const component = <DrupalPage mapper={mapper} url={'/'} />;

    expect(renderer.create(component).toJSON()).toMatchSnapshot();

    expect(
      renderer.create(await waitForHnData(component)).toJSON(),
    ).toMatchSnapshot();
  });

  test('with all props', async () => {
    const component = (
      <DrupalPage
        mapper={mapper}
        url={'/'}
        layout={'div'}
        layoutProps={{ testLayoutProp: true }}
        renderWhileLoadingData={true}
        pageProps={{ testPageProp: true }}
      />
    );

    expect(renderer.create(component).toJSON()).toMatchSnapshot();

    expect(
      renderer.create(await waitForHnData(component)).toJSON(),
    ).toMatchSnapshot();
  });
});
