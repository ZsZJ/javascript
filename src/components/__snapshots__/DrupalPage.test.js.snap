// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrupalPage with all props 1`] = `
<div
  loadingData={true}
  page={null}
  testLayoutProp={true}
  url={null}
/>
`;

exports[`DrupalPage with all props 2`] = `
<div
  loadingData={false}
  page={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle_1",
          "type": "unique_type_1",
        },
      },
    }
  }
  testLayoutProp={true}
  url="/"
>
  <div
    bundle="unique_type_1__unique_bundle_1"
    className="MyCustomMapperComponent"
    entity={
      Object {
        "__hn": Object {
          "entity": Object {
            "bundle": "unique_bundle_1",
            "type": "unique_type_1",
          },
        },
      }
    }
    page={
      Object {
        "__hn": Object {
          "entity": Object {
            "bundle": "unique_bundle_1",
            "type": "unique_type_1",
          },
        },
      }
    }
    paragraph={
      Object {
        "__hn": Object {
          "entity": Object {
            "bundle": "unique_bundle_1",
            "type": "unique_type_1",
          },
        },
      }
    }
    testPageProp={true}
  />
</div>
`;

exports[`DrupalPage with required props 1`] = `null`;

exports[`DrupalPage with required props 2`] = `
<div
  bundle="unique_type_1__unique_bundle_1"
  className="MyCustomMapperComponent"
  entity={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle_1",
          "type": "unique_type_1",
        },
      },
    }
  }
  page={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle_1",
          "type": "unique_type_1",
        },
      },
    }
  }
  paragraph={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle_1",
          "type": "unique_type_1",
        },
      },
    }
  }
/>
`;
